### PR TITLE
Adding Gamma AQ, HT, LT, and HW stress factors to biogenics

### LINKED ISSUE #124: [biogenics]  Add air quality/ozone stress gamma based on GFSv16
@MaggieMarvin   This should be relatively simple addition, where we can add the MEGAN3 function for calculating "GAMMA_AQ", which depends on a current air quality index, i.e. in MEGAN3 it uses ozone W126 (ppm hours-1).  This can be calculated based on input ozone concentrations.  Initially, we can make an option for a a set constant ozone mixing ratio input in the namelist and calculate the [W126](https://www.epa.gov/sites/default/files/2015-09/documents/w126_steps_to_calculate_revised_feb19.pdf) (easier) in canopy-app, but it should be spatially dependent for regional and global scale applications, thus calculated using gridded input ozone concentrations to canopy-app (trickier).

### PR DESCRIPTION
**Pull request by drnimbusrain**
This PR adds commits for the new biogenic stress factors including gamma aq (based on input or constant W126 ozone) and meteorological stress factors for extreme daily high/low temperatures (gamma_ht and gamma_lt) and for extreme high daily wind speeds (gamma_hw).  

These are based on MEGAN3...it seems in cases of very high ozone/W126, or in extreme temperatures (high/low) or wind speeds it can lead to increases in biogenic species emissions.   The details are in the README, where it is recommended that only the gammaht, gammalt, and gammaws are turned on only when using the historical option on with a simulation longer than 24 hours.

@MaggieMarvin  Please test all the options including gamma_aq, ht, lt, and hw with longer historical runs to gauge the increases in some biogenic emissions under extreme AQ or Met conditions.

### CODE REVIEW

**Comment by drnimbusrain** [comment]
image
@MaggieMarvin This result makes me think that the extreme met thresholds in MEGANv3 are not well constrained.  For instance, there is no dependence on species or PFT for these thresholds.  Clearly it doesn't work well for SESQ at the PECORINO sites (vegtype?).

Two ideas/options.  
1)  Add new NL parameters for user set optional thresholds and delta thresholds (e.g., w126_thresh and ht_thresh, dtht_thresh, lt_thresh, dtlt_thresh, hw_thresh, and dthw_thresh, see the [constant MEGAN3 values used here](https://github.com/noaa-oar-arl/canopy-app/blob/gamma_aq/src/canopy_bioparm_mod.F90#L514)) so that this can be tunable for any given site (maybe dependent on PFTs)....quicker solution.
2)  Explore literature about this effect to determine better PFT/species dependent extreme thresholds and add this dependence to canopy-app...longer solution.

**Review by MaggieMarvin — CHANGES_REQUESTED** [review]
From comparing against MEGANv3 code, it looks like CHW, CHT, and CLT of CARY should be set to 5 (as a sesquiterpene), and that CHW should be additionally set to 5 for all of the monoterpenes as well, including SABI, LIMO, CARE, BPIN, APIN, and MONO.

**Comment by drnimbusrain** [comment]
> From comparing against MEGANv3 code, it looks like CHW, CHT, and CLT of CARY should be set to 5 (as a sesquiterpene), and that CHW should be additionally set to 5 for all of the monoterpenes as well, including SABI, LIMO, CARE, BPIN, APIN, and MONO.

@MaggieMarvin Thank you for reviewing.  Did you make the suggested changes to the pertinent areas of the code in this review?

File: src/canopy_bioparm_mod.F90 [hunk 1]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: CHW_FARN         =  5.0_rk     !coefficient for high wind stress
+        REAL(RK),          PARAMETER     :: THW_FARN         =  12.0_rk    !threshold for high wind stress (m/s)
+        REAL(RK),          PARAMETER     :: DTHW_FARN        =  8.0_rk     !delta threshold for high wind stress (m/s)
+
         REAL(RK),          PARAMETER     :: CT1_CARY         =  130.0_rk   !Activation energy (kJ/mol)
         REAL(RK),          PARAMETER     :: CEO_CARY         =  2.37_rk    !Empirical coefficient
+        REAL(RK),          PARAMETER     :: CAQ_CARY         =  5.0_rk     !coefficient for poor Air Quality stress
+        REAL(RK),          PARAMETER     :: TAQ_CARY         =  20.0_rk    !threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: DTAQ_CARY        =  30.0_rk    !delta threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: CHT_CARY         =  1.0_rk     !coefficient for high temperature stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CHT_CARY         =  5.0_rk     !coefficient for high temperature stress
```

File: src/canopy_bioparm_mod.F90 [hunk 2]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: THT_CARY         =  313.15_rk  !threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTHT_CARY        =  8.0_rk     !delta threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: CLT_CARY         =  1.0_rk     !coefficient for low temperature stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CLT_CARY         =  5.0_rk     !coefficient for low temperature stress
```

File: src/canopy_bioparm_mod.F90 [hunk 3]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: TLT_CARY         =  283.15_rk  !threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTLT_CARY        =  8.0_rk     !delta threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: CHW_CARY         =  1.0_rk     !coefficient for high wind stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CHW_CARY         =  5.0_rk     !coefficient for high wind stress
```

File: src/canopy_bioparm_mod.F90 [hunk 4]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: TAQ_SABI         =  20.0_rk    !threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: DTAQ_SABI        =  30.0_rk    !delta threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: CHT_SABI         =  1.0_rk     !coefficient for high temperature stress
+        REAL(RK),          PARAMETER     :: THT_SABI         =  313.15_rk  !threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTHT_SABI        =  8.0_rk     !delta threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: CLT_SABI         =  1.0_rk     !coefficient for low temperature stress
+        REAL(RK),          PARAMETER     :: TLT_SABI         =  283.15_rk  !threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTLT_SABI        =  8.0_rk     !delta threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: CHW_SABI         =  1.0_rk     !coefficient for high wind stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CHW_SABI         =  5.0_rk     !coefficient for high wind stress
```

File: src/canopy_bioparm_mod.F90 [hunk 5]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: TAQ_LIMO         =  20.0_rk    !threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: DTAQ_LIMO        =  30.0_rk    !delta threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: CHT_LIMO         =  1.0_rk     !coefficient for high temperature stress
+        REAL(RK),          PARAMETER     :: THT_LIMO         =  313.15_rk  !threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTHT_LIMO        =  8.0_rk     !delta threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: CLT_LIMO         =  1.0_rk     !coefficient for low temperature stress
+        REAL(RK),          PARAMETER     :: TLT_LIMO         =  283.15_rk  !threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTLT_LIMO        =  8.0_rk     !delta threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: CHW_LIMO         =  1.0_rk     !coefficient for high wind stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CHW_LIMO         =  5.0_rk     !coefficient for high wind stress
```

File: src/canopy_bioparm_mod.F90 [hunk 6]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: TAQ_CARE         =  20.0_rk    !threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: DTAQ_CARE        =  30.0_rk    !delta threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: CHT_CARE         =  1.0_rk     !coefficient for high temperature stress
+        REAL(RK),          PARAMETER     :: THT_CARE         =  313.15_rk  !threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTHT_CARE        =  8.0_rk     !delta threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: CLT_CARE         =  1.0_rk     !coefficient for low temperature stress
+        REAL(RK),          PARAMETER     :: TLT_CARE         =  283.15_rk  !threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTLT_CARE        =  8.0_rk     !delta threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: CHW_CARE         =  1.0_rk     !coefficient for high wind stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CHW_CARE         =  5.0_rk     !coefficient for high wind stress
```

File: src/canopy_bioparm_mod.F90 [hunk 7]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: TAQ_BPIN         =  20.0_rk    !threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: DTAQ_BPIN        =  30.0_rk    !delta threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: CHT_BPIN         =  1.0_rk     !coefficient for high temperature stress
+        REAL(RK),          PARAMETER     :: THT_BPIN         =  313.15_rk  !threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTHT_BPIN        =  8.0_rk     !delta threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: CLT_BPIN         =  1.0_rk     !coefficient for low temperature stress
+        REAL(RK),          PARAMETER     :: TLT_BPIN         =  283.15_rk  !threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTLT_BPIN        =  8.0_rk     !delta threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: CHW_BPIN         =  1.0_rk     !coefficient for high wind stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CHW_BPIN         =  5.0_rk     !coefficient for high wind stress
```

File: src/canopy_bioparm_mod.F90 [hunk 8]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: TAQ_APIN         =  20.0_rk    !threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: DTAQ_APIN        =  30.0_rk    !delta threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: CHT_APIN         =  1.0_rk     !coefficient for high temperature stress
+        REAL(RK),          PARAMETER     :: THT_APIN         =  313.15_rk  !threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTHT_APIN        =  8.0_rk     !delta threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: CLT_APIN         =  1.0_rk     !coefficient for low temperature stress
+        REAL(RK),          PARAMETER     :: TLT_APIN         =  283.15_rk  !threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTLT_APIN        =  8.0_rk     !delta threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: CHW_APIN         =  1.0_rk     !coefficient for high wind stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CHW_APIN         =  5.0_rk     !coefficient for high wind stress
```

File: src/canopy_bioparm_mod.F90 [hunk 9]
```diff
@@ -496,45 +510,288 @@ SUBROUTINE CANOPY_BIOP( EMI_IND, LU_OPT, VTYPE, &
...
+        REAL(RK),          PARAMETER     :: TAQ_MONO         =  20.0_rk    !threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: DTAQ_MONO        =  30.0_rk    !delta threshold for poor Air Quality stress (ppm-hours)
+        REAL(RK),          PARAMETER     :: CHT_MONO         =  1.0_rk     !coefficient for high temperature stress
+        REAL(RK),          PARAMETER     :: THT_MONO         =  313.15_rk  !threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTHT_MONO        =  8.0_rk     !delta threshold for high temperature stress (K)
+        REAL(RK),          PARAMETER     :: CLT_MONO         =  1.0_rk     !coefficient for low temperature stress
+        REAL(RK),          PARAMETER     :: TLT_MONO         =  283.15_rk  !threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: DTLT_MONO        =  8.0_rk     !delta threshold for low temperature stress (K)
+        REAL(RK),          PARAMETER     :: CHW_MONO         =  1.0_rk     !coefficient for high wind stress
```
Comment (MaggieMarvin) [inline src/canopy_bioparm_mod.F90]:
```suggestion
        REAL(RK),          PARAMETER     :: CHW_MONO         =  5.0_rk     !coefficient for high wind stress
```

**Review by MaggieMarvin — COMMENTED** [review]
No, sorry! Changes are added here

**Comment by drnimbusrain** [comment]
> No, sorry! Changes are added here

Thank you for reviewing and the suggested changes.  I agree with you and have committed the suggestions to the branch.  Do you think we need more review and testing, or approve now, and improve later?  So far the results you show seem to make the emissions too high comparison to observations?  Truly, I do not yet really understand the relationship here.  

image

  For example, if the past 24-hr minimum temperature drops below threshold, it will boost emissions during the peak later in the day.  So far, I have only found a few mentions of how extreme temperatures, wind, or air quality could affect biogenic emissions.  Mainly on ozone impacts in the MEGAN 2006 paper, i.e.,"Large increases in the emission of isoprene have been observed from leaves exposed to short-term elevated ozone levels (Velikova et al., 2005). The response occurs rapidly and persists for hours after ozone levels are reduced."  This does somewhat correspond to how these factors work, where the daily max/min or AQ will have impacts that persist later into the day.  This suggests that larger diurnal fluctuations in the environment should lead to more pronounced emissions, but that's not what Pecorino shows.   
  
Also, do we have impacts of ozone on isoprene here?  I find really no documentation on these implementations in MEGANv3.   Do you find anything else on this?

**Comment by MaggieMarvin** [comment]
These effects do seem to be based on chamber experiments associated with the MEGAN development team (https://ui.adsabs.harvard.edu/abs/2018AGUFM.A33K3319N/abstract ; https://escholarship.org/uc/item/9wf58192), with evidence for a delay in monoterpene emissions (https://online.ucpress.edu/elementa/article/10/1/00096/122255/Impact-of-heat-stress-on-foliar-biogenic-volatile). The heat stress response is broadly supported by other recent studies (https://acp.copernicus.org/articles/23/2683/2023/), though you're probably correct that it really should be PFT-dependent (https://www.sciencedirect.com/science/article/pii/S2667010022001998). The influence of ozone on isoprene emissions seems unclear (ozone appears to reduce isoprene emissions in https://www.jstage.jst.go.jp/article/agrmet/73/4/73_D-17-00022/_pdf), but there is no response currently applied in the MEGAN3 implementation.

I believe that the code is working as intended - biogenic emissions of monoterpenes and sesquiterpenes increase with exposure to extreme stressors (see below for ozone W126 and AQ response for sesquiterpenes and min stress threshold of 20 ppm-hours) - but I'm not sure how broadly the thresholds and emission scaling factors should apply. I think it's ok to include for now since it's based on MEGAN3 and we can turn on/off (maybe set defaults to OFF?), but we should keep an eye out for MEGAN3 documentation and further advancements in the literature.

map_ozone_ppm-15
map_canopyapp_bio_global-2

**Comment by drnimbusrain** [comment]
@MaggieMarvin Thank you for these details, as they are very useful for future paper and for improving these (e.g., PFT dependence etc.) later if need be.  I have now changed all settings and descriptions for aq_opt, ht_opt, lt_opt, and hw_opt to Off by default.   Please go ahead and check.  Thank you.

After this, I think we have a comprehensive set of factors (with options) that impact and tune the biogenic emissions in canopy-app, and we can see if @quaz115  want's to (or has time to) work on updating the soil moisture factors I have there now.  Either way, we can move forward on your work to apply and evaluate canopy-app biogenics and paper!

File: src/canopy_readnml.F90 [hunk 10]
```diff
@@ -344,8 +344,8 @@ SUBROUTINE canopy_readnml
 !-------------------------------------------------------------------------------
 
 !------------------------------------------------------------------------------
-! Set default integer for using air quality stress gamma for biogenic emissions (default=0; On)
-    aq_opt = 0
+! Set default integer for using air quality stress gamma for biogenic emissions (default=1; Off)
```
Comment (MaggieMarvin) [inline src/canopy_readnml.F90]:
```suggestion
! Set default integer for using air quality stress gamma for biogenic emissions (default=2; Off)
```

**Review by MaggieMarvin — APPROVED** [review]
Looks great to me! Just one small inconsistency in canopy_readnml.F90 but otherwise I think it's ready to merge

**Pull request merged**